File: 2-medium/00002-add-two-numbers/solution.cpp
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        vector<short> n1{}, n2{}, n3{};
        ListNode* res =nullptr;
        while(l1 != nullptr) {
            n1.push_back(static_cast<short>(l1->val));
            l1 = l1->next;
        }
        while(l2 != nullptr) {
            n2.push_back(static_cast<short>(l2->val));
            l2 = l2->next;
        }
        short ofs{0};
        decltype(n1.size()) i{0}, j{0};
        while(true){
            short tmp = n1[i] + n2[j] + ofs;
            ofs = 0;
            if(tmp > 9) {
                tmp -= 10;
                ofs = 1;
            }
            n3.push_back(tmp);
            if(i==n1.size() -1 && j== n2.size() -1) {
                if(ofs > 0){
                    n3.push_back(ofs);
                }
                break;
            }
            if(i < n1.size() -1) {
                ++i;
            } else {
                n1[i] = 0;
            }
            if(j < n2.size() -1) {
                ++j;
            } else {
                n2[j] = 0;
            }
        }
        for(auto it{n3.rbegin()}; it != n3.rend(); ++it) {
            auto tmp = new ListNode();
            tmp->next = res;
            res=tmp;
            res->val = static_cast<int>(*it);
        }
        return res;
    }
};

File: 2-medium/00002-add-two-numbers/solution.rs
impl Solution {
    pub fn add_two_numbers(mut l1: Option<Box<ListNode>>, mut l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut n1, mut n2, mut n3) : (Vec<i8>, Vec<i8>, Vec<i8>) = (Vec::new(), Vec::new(), Vec::new());
        let mut res: Option<Box<ListNode>> = None;
        while let Some(ptr) = l1 {
            n1.push(ptr.val as i8);
            l1 = ptr.next;
        }
        while let Some(ptr) = l2 {
            n2.push(ptr.val as i8);
            l2 = ptr.next;
        }
        let (mut i, mut j, mut ofs) = (0usize, 0usize, 0i8);
        loop {
            let mut tmp = n1[i] + n2[j] + ofs;
            ofs = 0;
            if(tmp > 9) {
                tmp -= 10;
                ofs = 1;
            }
            n3.push(tmp);
            if i==n1.len() -1 && j== n2.len() -1 {
                if(ofs > 0){
                    n3.push(ofs);
                }
                break;
            }
            if i < n1.len() -1 {
                i += 1;
            } else {
                n1[i] = 0;
            }
            if j < n2.len() -1 {
                j += 1;
            } else {
                n2[j] = 0;
            }
        }
        for num in n3.iter().rev() {
            res = Some(Box::new(ListNode { next : res, val: *num as i32}));
        }
        res
    }
}
